server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java

```diff
@@ -230,376 +234,427 @@ private void onlyCreateIndex(final CreateIndexClusterStateUpdateRequest request,
+                                                             @Nullable IndexMetaData sourceMetaData) throws IOException {
+        MapperService mapperService = indexService.mapperService();
+        if (!mappings.isEmpty()) {
+            assert mappings.size() == 1 : mappings;
+            mapperService.merge(MapperService.SINGLE_MAPPING_NAME, mappings, MergeReason.MAPPING_UPDATE);
+        }
+
+        if (sourceMetaData == null) {
+            // now that the mapping is merged we can validate the index sort.
+            // we cannot validate for index shrinking since the mapping is empty
+            // at this point. The validation will take place later in the process
+            // (when all shards are copied in a single place).
+            indexService.getIndexSortSupplier().get();
+        }
+    }
+
+    private static IndexService validateActiveShardCountAndCreateIndexService(String indexName, ActiveShardCount waitForActiveShards,
+                                                                              Settings indexSettings, int routingNumShards,
+                                                                              IndicesService indicesService) throws IOException {
+        final IndexMetaData.Builder tmpImdBuilder = IndexMetaData.builder(indexName);
+        tmpImdBuilder.setRoutingNumShards(routingNumShards);
+        tmpImdBuilder.settings(indexSettings);
+
+        // Set up everything, now locally create the index to see that things are ok, and apply
+        IndexMetaData tmpImd = tmpImdBuilder.build();
+        if (waitForActiveShards == ActiveShardCount.DEFAULT) {
+            waitForActiveShards = tmpImd.getWaitForActiveShards();
         }
+        if (waitForActiveShards.validate(tmpImd.getNumberOfReplicas()) == false) {
+            throw new IllegalArgumentException("invalid wait_for_active_shards[" + waitForActiveShards +
+                "]: cannot be greater than number of shard copies [" +
+                (tmpImd.getNumberOfReplicas() + 1) + "]");
+        }
+        return indicesService.createIndex(tmpImd, Collections.emptyList());
     }
 
     private void validate(CreateIndexClusterStateUpdateRequest request, ClusterState state) {
         validateIndexName(request.index(), state);
         validateIndexSettings(request.index(), request.settings(), forbidPrivateIndexSettings);
     }
 
-    public void validateIndexSettings(String indexName, final Settings settings, final boolean forbidPrivateIndexSettings)
-            throws IndexCreationException {
+    public void validateIndexSettings(String indexName, Settings settings, boolean forbidPrivateIndexSettings)
+        throws IndexCreationException {
         List<String> validationErrors = getIndexSettingsValidationErrors(settings, forbidPrivateIndexSettings);
 
-        if (validationErrors.isEmpty() == false) {
+        if (!validationErrors.isEmpty()) {
```

Comment:
We prefer the other way because it's easier to see since a leading `!` is easier to accidentally gloss over

```suggestion
        if (validationErrors.isEmpty() == false) {
```